File: Source/NextTurnRuntime/Private/CoreUObject/PropertyClass.cpp
// Copyright (c) NextTurn. All rights reserved.
// Licensed under the Apache License, Version 2.0.
// See LICENSE.txt in the project root for more information.

#include "NextTurnRuntime/Private/Exports.hpp"

#include "NextTurnRuntime/Private/Accessor.hpp"
#include "NextTurnRuntime/Public/Compiler.hpp"

#include "Runtime/Core/Public/CoreMinimal.h"

ENABLE_WARNINGS;

DEFINE_OFFSET_ACCESSOR(FFieldClass, SuperClass);
DEFINE_OFFSET_ACCESSOR(FFieldClass, CastFlags);

#define TYPE_EXPORT_MANAGED_TYPE_NAME PropertyClass

TYPE_EXPORT_START

const FMemberSymbol Members[] =
{
	EXPORT_OFFSET(FFieldClass, BaseClass, SuperClass),
	EXPORT_OFFSET(FFieldClass, ClassTypeFlags, CastFlags),
};

TYPE_EXPORT_END

#undef TYPE_EXPORT_MANAGED_TYPE_NAME

RESTORE_WARNINGS;

File: Managed/NextTurn.UE.Runtime/CoreUObject/PropertyClass.cs
// Copyright (c) NextTurn. All rights reserved.
// Licensed under the Apache License, Version 2.0.
// See LICENSE.txt in the project root for more information.

using System;
using NextTurn.UE.Annotations;

namespace Unreal
{
    internal static class PropertyClass
    {
        internal static class NativeMethods
        {
            [ReadOffset]
            public static extern IntPtr GetBaseClass(IntPtr @class);

            [ReadOffset]
            public static extern ClassTypeFlags GetClassTypeFlags(IntPtr @class);

            public static bool HasAllTypeFlags(IntPtr @class, ClassTypeFlags flags) =>
                (GetClassTypeFlags(@class) & flags) == flags;
        }
    }
}
